tests/tests_unit/test_api/test_sequences.py

```diff
@@ -0,0 +1,256 @@
+        assert mock_seq_response.calls[0].response.json()["items"] == res.dump(camel_case=True)
+        req_body = jsgz_load(mock_seq_response.calls[0].request.body)
+        assert 42 == req_body["filter"]["lastUpdatedTime"]["max"]
+        assert {"name": "n", "description": "d", "query": "q"} == req_body["search"]
+
+    def test_update_object(self):
+        assert isinstance(
+            SequenceUpdate(1)
+            .asset_id.set(1)
+            .asset_id.set(None)
+            .description.set("")
+            .description.set(None)
+            .external_id.set("1")
+            .external_id.set(None)
+            .metadata.set({})
+            .metadata.add({})
+            .metadata.remove([])
+            .name.set("")
+            .name.set(None),
+            SequenceUpdate,
+        )
+
+    def test_insert(self, mock_post_datapoints):
+        data = {i: [2 * i] for i in range(1, 11)}
+        SEQ_API.data.insert(columns=[0], rows=data, external_id="eid")
+        assert {
+            "items": [
+                {
+                    "externalId": "eid",
+                    "columns": [{"id": 0}],
+                    "rows": [{"rowNumber": i, "values": [2 * i]} for i in range(1, 11)],
+                }
+            ]
+        } == jsgz_load(mock_post_datapoints.calls[0].request.body)
+
+    def test_retrieve_by_id(self, mock_get_datapoints):
+        data = SEQ_API.data.retrieve(id=123)
+        assert isinstance(data, list)
+        assert 1 == len(data)
+
+    def test_retrieve_by_external_id(self, mock_get_datapoints):
+        data = SEQ_API.data.retrieve(external_id="foo", start=1, end=1000)
+        assert isinstance(data, list)
+        assert 1 == len(data)
+
+    def test_delete_by_id(self, mock_delete_datapoints):
+        res = SEQ_API.data.delete(id=1, rows=[1, 2, 3])
+        assert res is None
+        assert {"items": [{"id": 1, "rows": [1, 2, 3]}]} == jsgz_load(mock_delete_datapoints.calls[0].request.body)
+
+    def test_delete_by_external_id(self, mock_delete_datapoints):
+        res = SEQ_API.data.delete(external_id="foo", rows=[1])
+        assert res is None
+        assert {"items": [{"externalId": "foo", "rows": [1]}]} == jsgz_load(
+            mock_delete_datapoints.calls[0].request.body
+        )
+
+
+@pytest.mark.dsl
+class TestSequencesPandasIntegration:
+    def test_retrieve_dataframe(self, mock_get_datapoints):
+        df = SEQ_API.data.retrieve_dataframe(external_id="foo", start=1000000, end=1100000)
+        assert {"ceid"} == set(df.columns)
+        assert df.shape[0] > 0
+        assert df.index == [0]
+
+    def test_retrieve_dataframe_convert_null(self, mock_get_datapoints_with_null):
+        df = SEQ_API.data.retrieve_dataframe(external_id="foo")
+        assert {"strcol", "intcol"} == set(df.columns)
+        assert "None" not in df.strcol
+        assert df.strcol.isna().any()
+        assert df.intcol.isna().any()
+        assert 2 == df.shape[0]
```

Comment:
Should test `.to_pandas()` on the Sequence and SequenceList objects aswell.

Reply:
done